### PR TITLE
[Feature] Update work experience UI

### LINKED ISSUE #11864: ✨ Update to work experience - UI changes
## ✨ Feature

We want to see a history of an employee's work in government for talent management. To this end, we want to collect group and level on the work experience. We also want to be able to match seniority in other orgs to those in gov, so we collect seniority when we do not collect group and level. Afterwards we will eventually deprecate collecting group and level in the "government employee" section.

## 🕵️ Details

There are a lot of things that show and hide, and display in various situations. Here's a list of all the changes and small interactions (but please also c[lick through the design](https://www.figma.com/proto/piRkRtWAHheoEX5Jq8d9u7/Create%2FEdit-an-experience-(Applicants)?node-id=487-58844&node-type=canvas&t=7ZVdFKtLK183qYms-0&scaling=min-zoom&content-scaling=fixed&page-id=472%3A70951)) :

### All experience
1. New question for all work experience "Employment category: External, Gov, CAF"
2. Start date and end date are moved to the bottom of form
3. Add a sentence to the description for "Highlight additional details" section stays at the bottom
4. In the designs the copy for "Current experience" is changed but honestly, we were just too lazy to revert all the designs back to match production. You do not need to change the labels for that checkbox.

### External
5. Add "size of org" question
6. Add "Seniority of the role" question (with "Self-employed" as an option)
7. **Read only state** - only needs to add those two fields, otherwise unchanged

### Gov
8. Instead of org, we ask for "Department, agency or other government organization" and it should be a Drop-down of the departments we have in the database.
9. Ask "Employment type"
10.  For Casual, term, or indeterminate we ask "Group" and "Level"
11. For indeterminate and term, we ask "Position type"
11.5 (**Gray added this late**) For term, we do **not** give the option for "substantive" in "Position type"
12. For Student, Casual, Term, Indeterminate-Acting, Indeterminate-Assignment, Indeterminate-Secondment, we ask for "Expected end date" if "Are you currently in this role?" is selected. After consultation with @petertgiles it was decided to store this data in the same field, just swap the label.
13. For Contractor we ask "Seniority of the role" (but no "Self-employed" option)
14. Contractor we ask "Contractor type" and if "firm or agency" is selected, we ask "Contracting firm or agency"
15. **Read only state** - The "Department, agency or other government organization" is shown where Organization normally shows up
16. **Read only state** - Add "Government of Canada" between Work experience and Dates
17. **Read only state** - Except if they are a Contractor. In that case replace "Government of Canada" with "Contractor"
18. **Read only state** - in the case of an end-date that's in the future. The text (expected end date) should be added
19. **Read only state** - Show employment type for all
20. **Read only state** - Show employment type, Position type and classification when collected

### CAF
21. Ask "Employment type" (Regular or Reserve)
22. Ask "Military force"
23. Ask Rank category
24. **Read only state** - The "Military force" ais shown where Organization normally shows up
25. **Read only state** - Add "Canadian Armed Forces" between Work experience and Dates
24. **Read only state** - Add "Employment type" and "Rank category" fields

## 🎨 Design
https://www.figma.com/design/piRkRtWAHheoEX5Jq8d9u7/Create%2FEdit-an-experience-(Applicants)?node-id=472-70951&t=NbjSJSxQjUSo26FF-1

## ✅ Acceptance Criteria

- [ ] Can view, add, and edit work experiences in career timeline
- [ ] Can view, add, and edit work experiences in application process
- [ ] Can view work experiences in admin profile view of career timeline
- [ ] Can view work experiences in admin view of snapshot - simple and new experience types
- [ ] Update backend validation to force entering new fields in the API
- [ ] Design review (Jerbo or Gray)

Out of scope: Classification picker with locking icon.  Use existing picker like on "create account" flow.
Out of scope: Downloadable files
Out of scope: Changing organization field to a searchable dropdown. It can stay as a text input.

## 🛑 Blockers

```[tasklist]
### Blocked By
- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/issues/11868
```

### PR DESCRIPTION
**Pull request by yonikid15**
🤖 Resolves #11864 

## 👋 Introduction

- Updates work experience form across site (Career timeline, application process)
- Updates the Experience cards across site to show new work experience fields. The cards vary depending on the employment category, and a few other fields. I advise using the [design examples](https://www.figma.com/proto/piRkRtWAHheoEX5Jq8d9u7/Create%2FEdit-an-experience-(Applicants)?node-id=487-58844&node-type=canvas&t=7ZVdFKtLK183qYms-0&scaling=min-zoom&content-scaling=fixed&page-id=472%3A70951) on the last slide, when reviewing the experience cards.
- Updates the Additional details sections help text

## 🕵️ Details

Please refer to the [notes](https://github.com/GCTC-NTGC/gc-digital-talent/issues/11864) in the issue, and the [designs](https://www.figma.com/proto/piRkRtWAHheoEX5Jq8d9u7/Create%2FEdit-an-experience-(Applicants)?node-id=487-58844&node-type=canvas&t=7ZVdFKtLK183qYms-0&scaling=min-zoom&content-scaling=fixed&page-id=472%3A70951).

NOTE: Still fixing the expected end date not getting added to the mutation.

## 🧪 Testing

1. Login as `admin@test.com`
2. Go to the Career Timeline page (/en/applicant/career-timeline). Ensure the add/edit form works correctly.
3. Create a new application. On experience step, ensure the add/edit form works correctly.
4. Ensure all the work experience cards are correct.

### CODE REVIEW

**Comment by petertgiles** [comment]
Could you please check on the tests?  Looks like there might be some updates to do there, too.

**Review by petertgiles — CHANGES_REQUESTED** [review]
I read through the code and didn't notice anything.  However, it seems like the form isn't initializing correctly for editing for me.

https://github.com/user-attachments/assets/0c0be92d-2252-4ab3-ac44-363b526a1e9d

**Review by esizer — CHANGES_REQUESTED** [review]
While you work on the tests, I noticed a few things.

**Comment by yonikid15** [comment]
> I read through the code and didn't notice anything. However, it seems like the form isn't initializing correctly for editing for me.

image


@petertgiles ah it's cause I just added the `defaultValue: null` to reset the sub fields when switching btw categories. I believe this should fix it.

File: apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  const experienceFormLabels = getExperienceFormLabels(intl);
+
+  return (
+    <>
+      <ContentSection
+        title={experienceFormLabels.team}
+        headingLevel={headingLevel}
+        data-h2-border-right="l-tablet(1px solid gray.lighter)"
+      >
+        {division ?? intl.formatMessage(commonMessages.notAvailable)}
```
Comment (vd1992) [inline apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx:26]: Division will always be null for CAF so it probably ought to not be mentioned or referenced at all on the card

image

File: apps/web/src/utils/experienceUtils.tsx [hunk 2]
```diff
@@ -260,6 +329,19 @@ export const formValuesToSubmitData = (
...
+    govContractorRoleSeniority,
+    govContractorType,
+    contractorFirmAgencyName,
+    classificationLevel: classificationId,
+    cafEmploymentType,
+    cafForce,
+    cafRank,
   } = data;
 
   const newEndDate = !currentRole && endDate ? endDate : null;
```
Comment (vd1992) [inline apps/web/src/utils/experienceUtils.tsx:347]: Will need to update this, given that currentRole is true and endDate is truthy is now a possibility with the expected end date

image

File: apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          intl.formatMessage(commonMessages.notAvailable)}
+      </ContentSection>
+      <Separator space="sm" />
+      <ContentSection
+        title={experienceFormLabels.cafRank}
+        headingLevel={headingLevel}
+        data-h2-border-right="l-tablet(1px solid gray.lighter)"
+      >
+        {getLocalizedName(cafRank?.label, intl) ??
+          intl.formatMessage(commonMessages.notAvailable)}
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx]: Fixed by [7ae5eb0](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/7ae5eb042526659eab491ba46aeda1a425c5b654)

File: apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  return (
+    <>
+      <ContentSection
+        title={experienceFormLabels.team}
+        headingLevel={headingLevel}
+        data-h2-border-right="l-tablet(1px solid gray.lighter)"
+      >
+        {division ?? intl.formatMessage(commonMessages.notAvailable)}
+      </ContentSection>
+      <Separator space="sm" />
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx]: Fixed by [8a4cfe1](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/8a4cfe1ecc278811be29ab7480a0ca79c2eb9af8)

File: apps/web/src/components/ExperienceCard/WorkContent/GovContent.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        >
+          {getLocalizedName(govEmploymentType.label, intl) ??
+            intl.formatMessage(commonMessages.notAvailable)}
+        </ContentSection>
+      </>
+    );
+  } else if (
+    govEmploymentType?.value === WorkExperienceGovEmployeeType.Casual ||
+    govEmploymentType?.value === WorkExperienceGovEmployeeType.Term
+  ) {
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceCard/WorkContent/GovContent.tsx:53]: Fixed by [9469226](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/9469226e586cd1fb0bd62ee0be6ab780891be097)

File: apps/web/src/components/ExperienceCard/WorkContent/GovContent.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+              {contractorFirmAgencyName ??
+                intl.formatMessage(commonMessages.notAvailable)}
+            </ContentSection>
+          </>
+        )}
+      </>
+    );
+  } else {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceCard/WorkContent/GovContent.tsx]: Fixed by [e3f680d](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/e3f680dda9847b69f5bf04e6e216be1f8f9943c3)

File: apps/web/src/components/ExperienceFormFields/WorkFields/CafFields.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  const intl = useIntl();
+  const [{ data, fetching }] = useQuery<CafFieldsOptionsQuery>({
+    query: CafFieldsOptions_Query,
+  });
+
+  const todayDate = new Date();
+  // to toggle whether End date is required, the state of the Current role checkbox must be monitored and have to adjust the form accordingly
+  const isCurrent = useWatch<{ currentRole: string }>({ name: "currentRole" });
+  // ensuring end date isn't before the start date, using this as a minimum value
+  const startDate = useWatch<{ startDate: string }>({ name: "startDate" });
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields/CafFields.tsx]: Fixed by [0c37760](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/0c377605f2a808b83c0789421800b27c6ccb895b)

File: apps/web/src/components/ExperienceFormFields/WorkFields/WorkFields.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  return (
+    <div>
+      {fetching ? (
+        <Loading inline />
+      ) : (
+        <div
+          data-h2-margin="base(x.5, 0, 0, 0)"
+          data-h2-max-width="base(50rem)"
+        >
+          <div data-h2-flex-grid="base(flex-start, x2, x1)">
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields/WorkFields.tsx:125]: Hmm, were using the same styling on some of the other Experiences as well. I think we should create a new ticket and do them all together. That way we can keep everything consistent.

File: apps/web/src/components/ExperienceFormFields/WorkFields/GovFields.tsx [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+    name: "govPositionType",
+  });
+  const watchGovContractorType = useWatch<{
+    govContractorType: GovContractorType;
+  }>({
+    name: "govContractorType",
+  });
+
+  const departmentOptions = unpackMaybes(data?.departments)
+    .filter(notEmpty)
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields/GovFields.tsx]: Removed in [965362e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/965362ea3cedeabbcd192bc37d30ed4bb963dc2e)

File: apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx (hunk 1 above)
Comment (yonikid15) [reply apps/web/src/components/ExperienceCard/WorkContent/CafContent.tsx]: Removed in [fc42f7a](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/fc42f7aae2d4af9aa87c470dc990f58a092b43f5)

File: apps/web/src/utils/experienceUtils.tsx (hunk 2 above)
Comment (yonikid15) [reply apps/web/src/utils/experienceUtils.tsx:350]: Thank you! I was looking for that! 
Updated endDate logic in [ec0e17b](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/ec0e17b104d4fb5bd6d0b993a9fb9f43b0f8e233)

File: apps/web/src/components/ExperienceFormFields/WorkFields/WorkFields.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (esizer) [reply apps/web/src/components/ExperienceFormFields/WorkFields/WorkFields.tsx:125]: oof, that is unfortunate :cry:

File: apps/web/src/components/ExperienceFormFields/WorkFields/CafFields.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (esizer) [reply apps/web/src/components/ExperienceFormFields/WorkFields/CafFields.tsx]: Okay, I was not aware `useWatch` has a different return type than `watch` so you needed to type cast it anyway. Still better I guess :sweat_smile:

File: apps/web/src/components/ExperienceFormFields/WorkFields.tsx
Comment (petertgiles) [inline apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: I think we're missing the option subtitles.
image

Comment (petertgiles) [inline apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: This is a bit sticky, but when you're editing an experience that had the "current position" option checked, it doesn't initialize that way on load.
image

Comment (petertgiles) [inline apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: I think there are some fields that aren't clearing when changing types.
image

Comment (petertgiles) [reply apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: When loading an existing record, can we assume "current position" is checked if end date is null or future?

File: api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php [hunk 10]
```diff
@@ -18,9 +19,9 @@ final class CreateUpdateWorkExperienceValidator extends Validator
     public function rules(): array
     {
         return [
-            // 'workExperience.employmentCategory' => [
-            //     'required',
-            // ],
+            'workExperience.employmentCategory' => [
+                'required',
+            ],
```
Comment (vd1992) [inline api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php:24]: Linking skills to an experience also uses this mutation, so the validator then trips up the Skills requirement step of an application. In that step the experience was already created, so you are editing it effectively.
Not sure what the approach should be here 🫤 

Also the success toast is misleading, it failed but that is a separate problem  

image

Comment (vd1992) [reply api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php:24]: The toast stems from below, the issue is `data` is truthy, just a null value connected to a key

image

Elsewhere we handle this by checking that `error` is null or querying into `data`

image

image

File: apps/web/src/pages/Applications/fragment.ts
Comment (vd1992) [inline apps/web/src/pages/Applications/fragment.ts:1]: Applications will need a comb through too
Like updating the naming here

image

**Review by vd1992 — CHANGES_REQUESTED** [review]
Big issue, maybe just getting larger 😅

File: api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php (hunk 10 above)
Comment (yonikid15) [reply api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php:24]: Nice catch :cry: I'm not sure where to from here. I think we can change the validation to use `sometimes`, so it [only validates it when present](https://laravel.com/docs/11.x/validation#validating-when-present). I'll check with @tristan-orourke.

File: apps/web/src/components/ExperienceFormFields/WorkFields.tsx
Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: Nice catch! Added in [595da1e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/595da1e7fece7ca53e1b36bc254fb795ef2cc8c4)

Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: Ya I believe that's the plan. Should be fixed in [5bdd920](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/5bdd920d35d88cd2160060dd364cef72a7d5f59f)

Comment (yonikid15) [reply apps/web/src/components/ExperienceFormFields/WorkFields.tsx:1]: Thanks for the help on this :+1: 
Fixed in [26d088e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/26d088e15fb111fa6cc80ce5d0299be5921a447c)

File: apps/web/src/pages/Applications/fragment.ts
Comment (yonikid15) [reply apps/web/src/pages/Applications/fragment.ts:1]: Nice catch vachan! Fixed in [4e94c97](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/4e94c97b543be3a9dc789b7f2db7df446cba282e)

File: api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php (hunk 10 above)
Comment (yonikid15) [reply api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php]: I tried it out in [276b5e2](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/276b5e210eeea535b18a9c8fadd29d01ebfee639) and it seems to be working.

Comment (yonikid15) [reply api/app/GraphQL/Validators/CreateUpdateWorkExperienceValidator.php]: I just chatted with Tristan. He said using the `sometimes` rule is fine for now. However, like Peter mentioned, we should make a ticket to create a separate mutation.

Rule added in [276b5e2](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/276b5e210eeea535b18a9c8fadd29d01ebfee639)

**Comment by yonikid15** [comment]
Reviewed with Jerbo and he approves :heavy_check_mark:

**Review by petertgiles — APPROVED** [review]

File: apps/playwright/tests/applicant/experience.spec.ts [hunk 11]
```diff
@@ -20,6 +22,115 @@ test.describe("Experiences", () => {
...
+  //   const workExperience = applicant.workExperiences?.find(
+  //     (ex) => ex?.role === role,
+  //   );
+
+  //   await experiencePage.editWorkExperience(`${workExperience?.id}`, {
+  //     role,
+  //     startDate: "2001-01",
+  //     endDate: "2200-01"
+  //   });
+  // });
```
Comment (esizer) [inline apps/playwright/tests/applicant/experience.spec.ts:132]: Was this intentional?

File: apps/web/src/utils/experienceUtils.tsx [hunk 12]
```diff
@@ -222,6 +232,68 @@ export const getExperienceFormLabels = (
...
+    positionType: intl.formatMessage({
+      defaultMessage: "Position type",
+      id: "0Dp1N4",
+      description: "Label for the position type radio group",
+    }),
+    govContractorRoleSeniority: intl.formatMessage({
+      defaultMessage: "Seniority of the role",
+      id: "WsG9lo",
+      description: "Label for the role seniority radio group",
+    }),
```
Comment (esizer) [inline apps/web/src/utils/experienceUtils.tsx:276]: This looks like a duplication of line 252. Are these contextually different for French? :thinking:

**Review by esizer — COMMENTED** [review]
I just have two questions right now.

File: apps/playwright/tests/applicant/experience.spec.ts (hunk 11 above)
Comment (yonikid15) [reply apps/playwright/tests/applicant/experience.spec.ts]: Nay, thanks for catching that :+1: Fixed in [b117b27](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12208/commits/b117b276c4e300e2bdc782c743fd406e0c086af4)

**Review by esizer — APPROVED** [review]
LGTM! :+1:

**Pull request merged**